### PR TITLE
Type mismatch pretty

### PR DESCRIPTION
**Pull request by Zuchos**
When no advice was applied to TypeMissmatchError clippy will make a console friendly diff on found and required types.

### CODE REVIEW

File: model/src/main/scala/com/softwaremill/clippy/StringDiff.scala [hunk 1]
```diff
@@ -0,0 +1,38 @@
+package com.softwaremill.clippy
+
+import com.softwaremill.clippy.StringDiff.{DELTA_END, DELTA_START}
+
+object StringDiff {
+  val ANSI_RESET = "\u001B[0m"
```
Comment (kciesielski) [inline model/src/main/scala/com/softwaremill/clippy/StringDiff.scala]: Scala convention for constants is `AnsiRed`, `DeltaEnd` etc.

File: model/src/main/scala/com/softwaremill/clippy/StringDiff.scala [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+  private def format(msg: String, expected: String, actual: String) = msg.format(expected, actual)
+
+  private def markDiff(source: String) = {
+    val prefix = findCommonPrefix()
+    val suffix = findCommonSuffix()
+    val diff = DELTA_START + source.substring(prefix.length, source.length - suffix.length) + DELTA_END
+    prefix + diff + suffix
+  }
+
+  private def findCommonPrefix(): String = {
```
Comment (kciesielski) [inline model/src/main/scala/com/softwaremill/clippy/StringDiff.scala]: This function can be reused so 

```scala
def findCommonSuffix() = 
   findCommonPrefix(expected.reverse, actual.reverse).reverse
```

File: model/src/main/scala/com/softwaremill/clippy/StringDiff.scala [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+object StringDiff {
+  val ANSI_RESET = "\u001B[0m"
+  val ANSI_RED = "\u001B[31m"
+  val DELTA_END = ANSI_RESET
+  val DELTA_START = ANSI_RED
+}
+
+class StringDiff(expected: String, actual: String) {
+
+
```
Comment (kciesielski) [inline model/src/main/scala/com/softwaremill/clippy/StringDiff.scala]: formatting: unnecessary gap

File: model/src/test/scala/com/softwaremill/clippy/StringDiffTest.scala [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+    ("Cool[String, String]", "Super[Option[String], String]", "expected " + DELTA_START + "Cool[String" + DELTA_END + ", String] but was " + DELTA_START + "Super[Option[String]" + DELTA_END + ", String]"),
+    ("(String, String)", "Super[Option[String], String]", "expected " + DELTA_START + "(String, String)" + DELTA_END + " but was " + DELTA_START + "Super[Option[String], String]" + DELTA_END)
+  )
+
+  "StringDiff" should "diff" in {
+    for ((expected, actual, expectedDiff) <- testData) {
+
+      val diff = new StringDiff(expected, actual).diff("expected %s but was %s")
+
+      diff should equal(expectedDiff)
```
Comment (kciesielski) [inline model/src/test/scala/com/softwaremill/clippy/StringDiffTest.scala]: By convention it's better to use `should equal` only when working with custom equality. Typically `should be` is preferred.

**Comment by kciesielski** [comment]
One more thing: Wouldn't things like `\u001B[31m` be displayed literally in, for example, Jenkins logs? It would make compilation errors actually harder to read :) Maybe this whole feature should be configurable by env variable, sth like running `sbt -Dclippy.colors=false test` in CI.

**Comment by ShaneDelmore** [comment]
It would be nice if it detected automatically if colors should be displayed, or if they were opt-in like they are with the scala 2.12 repl.  I agree that escape characters in ci logs would not be wanted and opt-out doesn't feel like the right default there.  I know the scalac devs have been kicking around ideas to determine if they should enable colors automatically.  Maybe you could detect if you are running in an interactive shell?

**Comment by ShaneDelmore** [comment]
I believe the escape codes also mess up emacs shell output.  I am not an emacs user myself but a couple of devs at my company who use clippy are and I would hate to mess up my messages for them.

**Comment by kciesielski** [comment]
@ShaneDelmore @Zuchos maybe coloring should be disabled by default, when enabling would require adding sth like
```scala
com.softwaremill.clippy.ClippySettings

clippyColors := true
```
wdyt?

**Comment by kciesielski** [comment]
@Zuchos Worth checking: https://github.com/lihaoyi/fansi, a "library for dealing with fancy colored Ansi strings easily, safely and efficiently". I'm thinking of using it here for coloring code in all errors, which will be a separate PR after you complete this one.

**Comment by kciesielski** [comment]
Or probably I'm going to need this :) http://www.lihaoyi.com/upickle-pprint/pprint/

**Comment by kciesielski** [comment]
@Zuchos @adamw @ShaneDelmore I added some updates to this PR. Now colors are disabled by default and can be enabled by adding `scalacOptions += "-P:clippy:colors=true"` compiler flag (can be set globally in `~/.sbt/0.13/clippy.sbt`).

File: README.md [hunk 5]
```diff
@@ -38,6 +38,15 @@ Upon first use, the plugin will download the advice dataset from `https://scala-
 `$HOME/.clippy` directory. The dataset will be updated at most once a day, in the background. You can customize the
 dataset URL and local store by using the `-P:clippy:url=` and `-P:clippy:store=` compiler options.
 
+# Enabling colors
+
+Clippy enhances some compiler errors with ANSI colors. If you'd like to enable colors on your terminal, you can
+`-P:clippy:colors=true` compiler option. To enable globally, add following line to `~/.sbt/0.13/clippy.sbt`:
```
Comment (adamw) [inline README.md]: "you can ... *add*" (missing I think). Also the "enable globally" applies to the case when using the SBT plugin, while the `-P` when using the compiler plugin directly

**Comment by kciesielski** [comment]
@adamw README and use page updated.

**Comment by adamw** [comment]
0.3.5 released

**Pull request merged**